### PR TITLE
Add sa token support for /metrics

### PR DESCRIPTION
**Pull request by rmweir**
**Problem:**
A service monitor pointing at rancher doesn't work with mounted service account token. This means a user needs to create an api token, then created a secret containing it.  Supporting the SA token will remove that process and also ensure that the used tokens expired since a rancher api token by default does not. Currently the API layer does not appear to support JWT tokens at all.

**Solution:**
Use a token review to auth tokens that have errored for /metrics path.

**Issue:**
https://github.com/rancher/rancher/issues/36234

### CODE REVIEW

File: pkg/auth/requests/token_review.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+		// This is to enable prometheus to use
+		// mounted service account tokens to
+		// fetch metrics.
+		"/metrics": true,
+	}
+)
+
+type TokenReviewAuth struct {
+	AuthClient           authv1.AuthenticationV1Interface
+	InitialAuthenticator auth.Authenticator
```
Comment (aiyengar2) [inline pkg/auth/requests/token_review.go:26]: I feel like the InitialAuthenticator call does not belong here. I think we should implement AuthenticatorFunc.Chain(authenticator Authenticator) [in rancher/steve](https://github.com/rancher/steve/blob/e7119828b8f9a9ca2cb2620ba650bd0508ca463e/pkg/auth/filter.go#L39-L43) or something similar to allow us to use something like 

```go
impersonateAuthenticator := requests.NewImpersonatingAuth(sar.NewSubjectAccessReview(clusterManager))
tokenAuthenticator := requests.NewTokenReviewAuth(scaledContext.K8sClient.AuthenticationV1())

authed.Use(mux.MiddlewareFunc(auth.ToMiddleware(impersonateAuthenticator.Chain(tokenAuthenticator))))
```

File: pkg/auth/requests/token_review.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+// Authenticate attempts to authenticate using given function. If authentication
+// fails AND the endpoint is in allowedPaths, it will attempt to perform a token
+// review to authenticate token and extract user info.
+func (t *TokenReviewAuth) Authenticate(req *http.Request) (user.Info, bool, error) {
+	info, hasAuth, err := t.InitialAuthenticator.Authenticate(req)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if info.GetName() != "system:cattle:error" {
```
[the remark this answers is not in the document]
Comment (aiyengar2) [reply pkg/auth/requests/token_review.go:45]: Also, I don't think this logic works since we don't parse a user to `system:cattle:error` during the Authenticate call; it happens after Authenticate is called in the Middleware function. Therefore, I think the code as it stands right now would never hit this point since it would have failed out with the original error.

I think this needs to be modified to:
```suggestion
	if hasAuth {
```

File: pkg/auth/requests/token_review.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		},
+		Spec: v1.TokenReviewSpec{
+			Token: tokens.GetTokenAuthFromRequest(req),
+		},
+	}
+
+	tokenReview, err = t.AuthClient.TokenReviews().Create(req.Context(), tokenReview, metav1.CreateOptions{})
+	if err != nil {
+		logrus.Debugf("tokenReview failed: %v", err)
+		return info, hasAuth, nil
```
Comment (aiyengar2) [inline pkg/auth/requests/token_review.go:68]:
```suggestion
		return info, false, nil
```

**Review by aiyengar2 — CHANGES_REQUESTED** [review]

File: pkg/auth/requests/token_review.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	if err != nil {
+		return nil, false, err
+	}
+
+	if info.GetName() != "system:cattle:error" {
+		// auth has succeeded
+		return info, hasAuth, nil
+	}
+
+	if !allowedPaths[req.URL.Path] {
```
Comment (cmurphy) [inline pkg/auth/requests/token_review.go:50]: Instead of checking paths here, why not just add this to only the /metrics route? https://github.com/rancher/rancher/pull/36220/files#diff-80627fe013488f5294c7e3c058e6231d1f4d622b3e7dd817bbf9db3c88af0cbbR105

File: pkg/auth/requests/token_review.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmweir) [reply pkg/auth/requests/token_review.go:45]: I think this happens because of the tokenAuthenticator that all requests go through. This code has been tested and when the token is invalid this is the name.

[the remark this answers is not in the document]
Comment (rmweir) [reply pkg/auth/requests/token_review.go:45]: Nothing gets set as a result of the initial Authenticate because it is not wrapped in the ToMiddleware func which is where that happens. So the info set is from a previous handler. I am changing this in my refactor to be more straight forward.

File: pkg/auth/requests/token_review.go (hunk 4 above)
Comment (rmweir) [reply pkg/auth/requests/token_review.go]: Changed to do just this.

File: pkg/auth/requests/token_review.go (hunk 1 above)
Comment (rmweir) [reply pkg/auth/requests/token_review.go:27]: I think the intended way to do this is to use toMiddleware on both and then chain them, I think I should have been doing that anyway since using that function ensure that the user info from the two functions are applied to the response in order. Fixed to do this.

File: pkg/auth/requests/token_review.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	authv1 "k8s.io/client-go/kubernetes/typed/authentication/v1"
+)
+
+var (
+	allowedPaths = map[string]bool{
+		// This is to enable prometheus to use
+		// mounted service account tokens to
+		// fetch metrics.
+		"/metrics": true,
+	}
```
[the remark this answers is not in the document]
Comment (rmweir) [reply pkg/auth/requests/token_review.go:22]: I used @cmurphy suggestion of just routing it to /metrics.

File: pkg/multiclustermanager/routes.go [hunk 6]
```diff
@@ -111,9 +110,21 @@ func router(ctx context.Context, localClusterEnabled bool, tunnelAuthorizer *mcm
 	authed.PathPrefix("/v3/token").Handler(tokenAPI)
 	authed.PathPrefix("/v3").Handler(managementAPI)
 
+	// Authenticated routes
+	metricsAuthed := mux.NewRouter()
+	metricsAuthed.UseEncodedPath()
+	impersonsatingAuth := auth.ToMiddleware(requests.NewImpersonatingAuth(sar.NewSubjectAccessReview(clusterManager)))
```
Comment (cmurphy) [inline pkg/multiclustermanager/routes.go:116]:
```suggestion
	impersonatingAuth := auth.ToMiddleware(requests.NewImpersonatingAuth(sar.NewSubjectAccessReview(clusterManager)))
```

File: pkg/multiclustermanager/routes.go [hunk 7]
```diff
@@ -102,7 +102,6 @@ func router(ctx context.Context, localClusterEnabled bool, tunnelAuthorizer *mcm
 	authed.Path("/meta/oci/{resource}").Handler(oci.NewOCIHandler(scaledContext))
 	authed.Path("/meta/vsphere/{field}").Handler(vsphere.NewVsphereHandler(scaledContext))
 	authed.Path("/v3/tokenreview").Methods(http.MethodPost).Handler(&webhook.TokenReviewer{})
-	authed.Path("/metrics").Handler(metricsHandler)
 	authed.Path("/metrics/{clusterID}").Handler(metricsHandler)
```
Comment (cmurphy) [inline pkg/multiclustermanager/routes.go:105]: This would need to apply to this route too right?

File: pkg/multiclustermanager/routes.go [hunk 8]
```diff
@@ -111,9 +110,21 @@ func router(ctx context.Context, localClusterEnabled bool, tunnelAuthorizer *mcm
...
 	authed.PathPrefix("/v3").Handler(managementAPI)
 
+	// Authenticated routes
+	metricsAuthed := mux.NewRouter()
+	metricsAuthed.UseEncodedPath()
+	impersonsatingAuth := auth.ToMiddleware(requests.NewImpersonatingAuth(sar.NewSubjectAccessReview(clusterManager)))
+	tokenReviewAuth := impersonsatingAuth.Chain(auth.ToMiddleware(requests.NewTokenReviewAuth(scaledContext.K8sClient.AuthenticationV1())))
+	metricsAuthed.Use(mux.MiddlewareFunc(tokenReviewAuth))
+	metricsAuthed.Use(mux.MiddlewareFunc(rbac.NewAccessControlHandler()))
+	metricsAuthed.Use(requests.NewAuthenticatedFilter)
```
Comment (cmurphy) [inline pkg/multiclustermanager/routes.go:120]: I was hoping there would be an easier way to add a middleware to one route without this duplicate code but it looks like I was mistaken

File: pkg/multiclustermanager/routes.go [hunk 9]
```diff
@@ -111,9 +110,21 @@ func router(ctx context.Context, localClusterEnabled bool, tunnelAuthorizer *mcm
...
+
+	metricsAuthed.Path("/metrics").Handler(metricsHandler)
+
 	unauthed.NotFoundHandler = authed
 	return func(next http.Handler) http.Handler {
-		authed.NotFoundHandler = next
+		authed.NotFoundHandler = metricsAuthed
```
Comment (cmurphy) [inline pkg/multiclustermanager/routes.go:126]: Why is this needed?

Comment (rmweir) [reply pkg/multiclustermanager/routes.go:126]: This chains the routing logic together, first it looks for routes from the unauth handler, if nothing matches then from auth, if nothing matches then from metrics auth. This is also relates to your two above questions. If metrics matches in the auth handler it will not be forwarded to the metrics handler. Using a new auth handler is how we are only using this auth for metrics otherwise we could use less code and just use the auth handler with metrics endpoint but then that performs the metrics auth logic on everything that fails first auth check.

File: pkg/multiclustermanager/routes.go [hunk 10]
```diff
@@ -111,9 +110,21 @@ func router(ctx context.Context, localClusterEnabled bool, tunnelAuthorizer *mcm
 	authed.PathPrefix("/v3/token").Handler(tokenAPI)
 	authed.PathPrefix("/v3").Handler(managementAPI)
 
+	// Authenticated routes
```
[the remark this answers is not in the document]
Comment (rmweir) [reply pkg/multiclustermanager/routes.go]: Changed to "Metrics authenticated route".

**Review by cmurphy — APPROVED** [review]

File: pkg/multiclustermanager/routes.go (hunk 9 above)
Comment (aiyengar2) [reply pkg/multiclustermanager/routes.go:126]: Since this line does not use the `next` var, can we move it outside of this function (just like how we set `unauthed.NotFoundHandler`? I think it only needs to be set once.

File: pkg/multiclustermanager/routes.go [hunk 11]
```diff
@@ -111,9 +110,21 @@ func router(ctx context.Context, localClusterEnabled bool, tunnelAuthorizer *mcm
 	authed.PathPrefix("/v3/token").Handler(tokenAPI)
 	authed.PathPrefix("/v3").Handler(managementAPI)
 
+	// Metrics authenticated route
+	metricsAuthed := mux.NewRouter()
+	metricsAuthed.UseEncodedPath()
+	impersonatingAuth := auth.ToMiddleware(requests.NewImpersonatingAuth(sar.NewSubjectAccessReview(clusterManager)))
+	tokenReviewAuth := impersonatingAuth.Chain(auth.ToMiddleware(requests.NewTokenReviewAuth(scaledContext.K8sClient.AuthenticationV1())))
+	metricsAuthed.Use(mux.MiddlewareFunc(tokenReviewAuth))
+	metricsAuthed.Use(mux.MiddlewareFunc(rbac.NewAccessControlHandler()))
```
Comment (aiyengar2) [inline pkg/multiclustermanager/routes.go:119]: Can we reuse the `rbac.NewAccessControlHandler()` from above instead of calling it independently?

File: pkg/multiclustermanager/routes.go [hunk 12]
```diff
@@ -111,9 +110,21 @@ func router(ctx context.Context, localClusterEnabled bool, tunnelAuthorizer *mcm
 	authed.PathPrefix("/v3/token").Handler(tokenAPI)
 	authed.PathPrefix("/v3").Handler(managementAPI)
 
+	// Metrics authenticated route
+	metricsAuthed := mux.NewRouter()
+	metricsAuthed.UseEncodedPath()
+	impersonatingAuth := auth.ToMiddleware(requests.NewImpersonatingAuth(sar.NewSubjectAccessReview(clusterManager)))
```
Comment (aiyengar2) [inline pkg/multiclustermanager/routes.go:116]: can we define this above and use this in https://github.com/rancher/rancher/blob/978e165156362d98a323452049fb04c7fec64f5d/pkg/multiclustermanager/routes.go#L96 as well?

Based on the fact that the logic for [`Chain`](https://github.com/rancher/steve/blob/e7119828b8f9a9ca2cb2620ba650bd0508ca463e/pkg/auth/filter.go#L47-L51) emits a brand new function, we should be able to reuse the same function for impersonating auth in both places rather than defining it twice.

File: pkg/multiclustermanager/routes.go [hunk 13]
```diff
@@ -111,9 +110,21 @@ func router(ctx context.Context, localClusterEnabled bool, tunnelAuthorizer *mcm
...
+	tokenReviewAuth := impersonatingAuth.Chain(auth.ToMiddleware(requests.NewTokenReviewAuth(scaledContext.K8sClient.AuthenticationV1())))
```
Comment (aiyengar2) [inline pkg/multiclustermanager/routes.go:117]: Is this the right order? I'm not sure if this is correct or if the following is correct:

```go
auth.ToMiddleware(requests.NewTokenReviewAuth(scaledContext.K8sClient.AuthenticationV1())).Chain(impersonatingAuth)
```

Based on [the logic in Steve](https://github.com/rancher/steve/blob/e7119828b8f9a9ca2cb2620ba650bd0508ca463e/pkg/auth/filter.go#L47-L51) that looks like `return m(middleware(next))`, it seems like we evaluate the `middleware` provided as an argument (e.g. the tokenAuthReview) before evaluating the `m` provided as self (e.g. the impersonatingAuth), whereas what we want is for impersonation to be attempted, followed by tokenAuthReview to be checked.

File: pkg/auth/requests/token_review.go [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+	if err != nil {
+		logrus.Debugf("tokenReview failed: %v", err)
+		return info, false, nil
+	}
+
+	tokenReviewUserInfo := &user.DefaultInfo{
+		Name:   tokenReview.Status.User.Username,
+		UID:    tokenReview.Status.User.UID,
+		Groups: tokenReview.Status.User.Groups,
+	}
```
Comment (aiyengar2) [inline pkg/auth/requests/token_review.go:56]: Do we need to do some check to wait till the status is populated?

**Review by aiyengar2 — CHANGES_REQUESTED** [review]

File: pkg/auth/requests/token_review.go (hunk 14 above)
Comment (rmweir) [reply pkg/auth/requests/token_review.go:56]: I believe the status should always be populated https://kubernetes.io/docs/reference/access-authn-authz/authentication/#webhook-token-authentication:
`The remote service is expected to fill the status field of the request to indicate the success of the login.`

File: pkg/multiclustermanager/routes.go (hunk 13 above)
Comment (rmweir) [reply pkg/multiclustermanager/routes.go]: Effectively I think either way works but you're right this isn't the intended order, changed.

File: pkg/multiclustermanager/routes.go (hunk 8 above)
Comment (rmweir) [reply pkg/multiclustermanager/routes.go:122]: Thanks to some of Arvind's feedback I was able to reuse certain parts.

File: pkg/multiclustermanager/routes.go (hunk 7 above)
Comment (cmurphy) [reply pkg/multiclustermanager/routes.go:108]: What about this?

**Review by cmurphy — CHANGES_REQUESTED** [review]

**Comment by rmweir** [comment]
@cmurphy I explained this here https://github.com/rancher/rancher/pull/36220#discussion_r808543272. Basically the metrics endpoint is fully covered by the second router and the request will not be forwarded to that router unless the route is not found in the regular auth router; this is what's going on in the `NotFoundHandler` part.

**Comment by cmurphy** [comment]
@rmweir in that case shouldn't the `/metrics/{clusterID}` route be removed from the regular authed router?

**Review by aiyengar2 — DISMISSED** [review]
LGTM

**Comment by cmurphy** [comment]
> @rmweir in that case shouldn't the /metrics/{clusterID} route be removed from the regular authed router?

I ran this through a debugger and found that querying /metrics/{clusterID} doesn't hit this token review authenticator. I assumed you want both /metrics and /metrics/{clusterID} to be accessible by service accounts, is that incorrect?

**Review by cmurphy — DISMISSED** [review]
Got it, thanks!

**Review by cbron — APPROVED** [review]
